libethcore/Transaction.cpp

```diff
@@ -64,9 +65,15 @@ TransactionBase::TransactionBase(bytesConstRef _rlpData, CheckTransaction _check
 			BOOST_THROW_EXCEPTION(InvalidTransactionFormat() << errinfo_comment("transaction data RLP must be an array"));
 
 		m_data = rlp[field = 5].toBytes();
-		byte v = rlp[field = 6].toInt<byte>() - 27;
+		m_chainID = _chainID;
+
+		byte v = rlp[field = 6].toInt<byte>();
 		h256 r = rlp[field = 7].toInt<u256>();
 		h256 s = rlp[field = 8].toInt<u256>();
+		if(v == m_chainID*2+35 || v == m_chainID*2+36)
```

Comment:
Here you should calculate the transaction chain id as `txChainId = (v - 35) / 2`. Then `txChainId` either match `m_chainId` or is `-4` for Frontier transactions. If it is other value you should throw an exception to reject the transaction.